### PR TITLE
NIFI-12277: Add SSLContextService to Slack processors

### PR DESCRIPTION
**Pull request by mark-bathori**
# Summary

[NIFI-12277](https://issues.apache.org/jira/browse/NIFI-12277)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PostSlack.java [hunk 1]
```diff
@@ -180,6 +188,13 @@ public class PostSlack extends AbstractProcessor {
             .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
             .build();
 
+    public static final PropertyDescriptor SSL_CONTEXT_SERVICE = new PropertyDescriptor.Builder()
+            .name("SSL Context Service")
+            .description("If specified, indicates the SSL Context Service that is used to communicate with the "
+                    + "remote server. If not specified, communications will not be encrypted")
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PostSlack.java:194]: This description is not correct, in absence of the `SSL Context Service`, communication will be encrypted, but the Processor will use the system default certificate authorities.

File: nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java [hunk 2]
```diff
@@ -257,16 +272,27 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
             final URL url = URI.create(context.getProperty(WEBHOOK_URL).evaluateAttributeExpressions(flowFile).getValue()).toURL();
-            final HttpURLConnection conn = (HttpURLConnection) url.openConnection();
-            conn.setRequestMethod("POST");
-            conn.setDoOutput(true);
+            int responseCode;
+
+            if (sslService != null) {
+                final HttpsURLConnection conn = (HttpsURLConnection) url.openConnection();
+                conn.setRequestMethod("POST");
+                conn.setDoOutput(true);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java:281]: These lines are the same for both conditionals, so recommend restructuring the approach to remove the duplication.

File: nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java [hunk 3]
```diff
@@ -257,16 +272,27 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
             jsonWriter.close();
 
             final URL url = URI.create(context.getProperty(WEBHOOK_URL).evaluateAttributeExpressions(flowFile).getValue()).toURL();
-            final HttpURLConnection conn = (HttpURLConnection) url.openConnection();
-            conn.setRequestMethod("POST");
-            conn.setDoOutput(true);
+            int responseCode;
+
+            if (sslService != null) {
+                final HttpsURLConnection conn = (HttpsURLConnection) url.openConnection();
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java:278]: `HttpsURLConnection` is a subclass of `HttpURLConnection`, so the approach should be restructured to remove the duplicative elements and scope the condition to setting the SSLSocketFactory when the SSL Context Service is configured.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for working on this new capability @mark-bathori. Although in general, an SSL Context Service should not be necessary for communicating with Slack because it has certificates signed by standard certificate authorities, this option can be useful in some network environments.

I noted a concern regarding the property description, and also recommended some refactoring to reduce duplication in `PutSlack`.

**Comment by mark-bathori** [comment]
Thanks for the review @exceptionfactory. Yes, the ticket’s main goal is to provide a possibility to use custom certificates when the system default certificates are not suitable.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for making the adjustments @mark-bathori.

Unfortunately it appears that the tests are failing repeatedly on the Windows builds.

Although unit test changes is a good best practice, this might be a case where it is less useful, given the expense of creating temporary keys and files. I would consider removing the new test methods.

```
Error:  org.apache.nifi.processors.slack.PostSlackTextMessageTest.sendTextOnlyMessageWithSSLContext -- Time elapsed: 4.141 s <<< FAILURE!
org.opentest4j.AssertionFailedError: Expected all Transferred FlowFiles to go to success but 1 were routed to failure
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:38)
	at org.junit.jupiter.api.Assertions.fail(Assertions.java:134)
	at org.apache.nifi.util.MockProcessSession.assertAllFlowFilesTransferred(MockProcessSession.java:1301)
	at org.apache.nifi.util.StandardProcessorTestRunner.assertAllFlowFilesTransferred(StandardProcessorTestRunner.java:319)
	at org.apache.nifi.processors.slack.PostSlackTextMessageTest.sendTextOnlyMessageWithSSLContext(PostSlackTextMessageTest.java:301)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)

Error:  Tests run: 13, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.237 s <<< FAILURE! -- in org.apache.nifi.processors.slack.PutSlackTest
Error:  org.apache.nifi.processors.slack.PutSlackTest.testWithSSLContext -- Time elapsed: 1.017 s <<< FAILURE!
org.opentest4j.AssertionFailedError: Expected all Transferred FlowFiles to go to success but 1 were routed to failure
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:38)
	at org.junit.jupiter.api.Assertions.fail(Assertions.java:134)
	at org.apache.nifi.util.MockProcessSession.assertAllFlowFilesTransferred(MockProcessSession.java:1301)
	at org.apache.nifi.util.StandardProcessorTestRunner.assertAllFlowFilesTransferred(StandardProcessorTestRunner.java:319)
	at org.apache.nifi.util.StandardProcessorTestRunner.assertAllFlowFilesTransferred(StandardProcessorTestRunner.java:398)
	at org.apache.nifi.processors.slack.PutSlackTest.testWithSSLContext(PutSlackTest.java:269)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

**Comment by mark-bathori** [comment]
Agree on removing the tests, pushed the changes.

**Review by exceptionfactory — APPROVED** [review]
Thanks for working through the feedback and making the adjustments @mark-bathori, the latest version looks good! +1 merging

**Pull request closed without merging**